robocorp-code/vscode-client/src/pythonExtIntegration.ts

```diff
@@ -43,26 +60,77 @@ export async function autoUpdateInterpreter(docUri: Uri) {
             ConfigurationTarget.WorkspaceFolder
         );
     }
+    return true;
 }
 
-export async function installPythonInterpreterCheck(context: ExtensionContext) {
-    context.subscriptions.push(
-        window.onDidChangeActiveTextEditor(async (event: TextEditor) => {
-            try {
-                // Whenever the active editor changes we update the Python interpreter used (if needed).
-                let docUri = event?.document?.uri;
-                if (docUri) {
-                    await autoUpdateInterpreter(docUri);
+export async function installWorkspaceWatcher(context: ExtensionContext) {
+    // listen to editor change/switch (this should cause environment switching as well)
+    const checkEditorSwitch = window.onDidChangeActiveTextEditor(async (event: TextEditor) => {
+        try {
+            // Whenever the active editor changes we update the Python interpreter used (if needed).
+            let docURI = event?.document?.uri;
+            if (docURI) {
+                await autoUpdateInterpreter(docURI);
+            }
+        } catch (error) {
+            logError("Error auto-updating Python interpreter.", error, "PYTHON_SET_INTERPRETER");
+        }
+    });
+
+    // listen for document changes and mark targeted files as dirty
+    const checkIfFilesHaveChanged = workspace.onDidChangeTextDocument(async (event) => {
+        let docURI = event.document.uri;
+        if (event.document.isDirty && (isEnvironmentFile(docURI.fsPath) || isPythonFile(docURI.fsPath))) {
+            dirtyWorkspaceFiles.add(docURI.fsPath);
+        }
+    });
+
+    // listen for when documents are about to be saved
+    const checkIfFilesWillBeSaved = workspace.onDidSaveTextDocument(async (document) => {
+        try {
+            let docURI = document.uri;
+            if (
+                docURI &&
+                dirtyWorkspaceFiles.has(docURI.fsPath) &&
```

Comment:
I'm not sure I follow the need for `dirtyWorkspaceFiles`... isn't just listening for the save and checking `isEnvironmentFile(docURI.fsPath) || isPythonFile(docURI.fsPath)` enough? 

I mean the save should only happen when an actual file was saved (in which case it must've changed in the meanwhile, no?) -- if that's the case the

Reply:
One can still trigger the save listener even if the file hasn't been modified. That's why I've added the dirty files store so that we trigger only when the file has been modified.